chia/data_layer/data_layer.py

```diff
@@ -136,3 +141,137 @@ async def get_root(self, store_id: bytes32) -> Optional[bytes32]:
+                tree_id, record.root, max_generation
+            )
+            if root is None or root.generation < min_generation:
+                return False
+
+            self.log.info(f"Validated chain hash {record.root} in downloaded datastore.")
+            max_generation = root.generation
+            last_checked_hash = record.root
+
+        return True
+
+    async def fetch_and_validate(self, subscription: Subscription) -> None:
+        tree_id = subscription.tree_id
+        singleton_record: Optional[SingletonRecord] = await self.wallet_rpc.dl_latest_singleton(tree_id)
+        if singleton_record is None:
+            return
+        if singleton_record.generation == uint32(0):
+            return
+        old_root: Optional[Root] = None
+        if await self.data_store.tree_id_exists(tree_id=tree_id):
+            old_root = await self.data_store.get_tree_root(tree_id=tree_id)
+        wallet_current_generation = await self.data_store.get_wallet_generation(tree_id)
+        assert int(wallet_current_generation) <= singleton_record.generation
+        # Wallet generation didn't change, so no new data committed on chain.
+        if wallet_current_generation is not None and uint32(wallet_current_generation) == singleton_record.generation:
+            return
+        to_check: List[SingletonRecord] = []
+        if subscription.mode is DownloadMode.LATEST:
+            to_check = [singleton_record]
+        if subscription.mode is DownloadMode.HISTORY:
+            to_check = await self.wallet_rpc.dl_history(
+                launcher_id=tree_id, min_generation=uint32(wallet_current_generation + 1)
+            )
+        # No root hash changes in the new wallet records, so ignore.
+        if (
+            old_root is not None
+            and old_root.node_hash is not None
+            and to_check[0].root == old_root.node_hash
+            and len(set(record.root for record in to_check)) == 1
+        ):
+            await self.data_store.set_wallet_generation(tree_id, int(singleton_record.generation))
+            return
+        # Delete all identical root hashes to our old root hash, until we detect a change.
+        if old_root is not None and old_root.node_hash is not None:
+            while to_check[-1].root == old_root.node_hash:
+                to_check.pop()
+
+        self.log.info(
+            f"Downloading and validating {subscription.tree_id}. "
+            f"Current wallet generation: {int(wallet_current_generation)}. "
+            f"Target wallet generation: {singleton_record.generation}."
+        )
+
+        downloaded = await self.data_store.download_data(subscription, singleton_record.root)
+        if not downloaded:
+            raise RuntimeError("Could not download the data.")
+
+        root = await self.data_store.get_tree_root(tree_id=tree_id)
+        # Wallet root hash must match to our data store root hash.
+        if root.node_hash is None or root.node_hash != to_check[0].root:
```

Comment:
is the comment wrong ?
it looks like both these root hashes where retrieved from the data_store and not the wallet

Reply:
to_check is retrieved from dl_wallet:

```
singleton_record: Optional[SingletonRecord] = await self.wallet_rpc.dl_latest_singleton(tree_id)
if subscription.mode is DownloadMode.LATEST:
    to_check = [singleton_record]
if subscription.mode is DownloadMode.HISTORY:
    to_check = await self.wallet_rpc.dl_history(
        launcher_id=tree_id, min_generation=uint32(wallet_current_generation + 1)
    )
```

Reply:
sorry i got a bit confused there